lib/chef/provider/package/zypper.rb

```diff
@@ -70,7 +72,22 @@ def get_current_versions
         end
 
         def candidate_version
-          @candidate_version ||= package_name_array.each_with_index.map { |pkg, i| available_version(i) }
+          if new_resource.source
+            logger.trace("#{new_resource} checking rpm status")
+            resolve_source_to_version
```

Comment:
this won't work because you're passing back one of the Version objects into the candidate_version.

if you look at the yum package provider this code goes into the `available_version` helper and then that is rendered into an string by the loop which is in `candidate_version` there.  if we go down the road of adding the Version object then that pattern needs to be adhered to.

this also calls out the fact that this code needs functional tests, because this code as it stands is pretty certainly broken but all the unit tests pass.  so we need code in spec/functional/resource/zypper_package_spec.rb which does a real install of a source package (i think there's examples in the yum/dnf functional tests a well, and the zypper func tests are written similarly so it should be easy enough to copy and get a working functional test).